Implement RunAsGroup Strategy in PSP

krmayankk opened the pull request:
```release-note
Add ability to control primary GID of containers through Pod Spec and PodSecurityPolicy
```
Part of feature  https://github.com/kubernetes/features/issues/213

Commit:
api changes for psp runasgroup policy

krmayankk commented:
/assign @tallclair

krmayankk commented:
/assign @liggitt

wenjiaswe commented:
@jennybuckley

enj commented:
@kubernetes/sig-auth-api-reviews

jennybuckley commented:
Looks like this PR was added to sig/api-machinery automatically based on the files it touches, but it is probably more applicable to the other sigs tagged here, since it's just making changes to a specific API

/remove-sig api-machinery

krmayankk commented:
/test pull-kubernetes-e2e-gce

krmayankk commented:
/test pull-kubernetes-e2e-gce

krmayankk commented:
/test pull-kubernetes-e2e-gce

liggitt commented on pkg/apis/policy/types.go:
```diff
@@ -228,6 +228,8 @@ type PodSecurityPolicySpec struct {
 	// +optional
 	ForbiddenSysctls []string
+	// RunAsGroup is the strategy that will dictate the allowable RunAsGroup values that may be set.
+	RunAsGroup RunAsGroupStrategyOptions
```
move to just after `RunAsUser`

liggitt replied:
document this is alpha-level and requires the RunAsGroup feature gate to be enabled

liggitt commented on pkg/apis/policy/types.go:
```diff
@@ -352,6 +364,19 @@ const (
+	// RunAsGroupStrategyMustRunAs means that container must run as a particular uid.
+	RunAsGroupStrategyMustRunAs RunAsGroupStrategy = "MustRunAs"
+	// RunAsGroupStrategyMustRunAsNonRoot means that container must run as a non-root uid
+	RunAsGroupStrategyMustRunAsNonRoot RunAsGroupStrategy = "MustRunAsNonRoot"
```
drop this one until https://github.com/kubernetes/kubernetes/pull/62216 is decided on?

tallclair replied:
Do we even need this? Isn't it the same as:
```yaml
runAsGroup:
  rule: MustRunAs
  ranges:
    - min: 1
      max: 2147483647
```

krmayankk replied:
@liggitt   i think #62216  is mostly about not needing NonRootGroup in Pod Spec. I think there is agreement that we need this feature at PSP level.
@tallclair agree MustRunAs can achieve that. It would be just inconsistent with how we do non root for RunAsUser.

its basically a choice between what you posted vs this below. More about usability. I think just from usability and inconsistency point of view we should stick with the bool
```
runAsGroup:
  runAsNonRoot: true
```

liggitt replied:
Until we have a dedicated runAsNonRootGroup field in the pod spec, I'd prefer to model the PSP restriction as Tim described above. The only reason we have a runAsNonRoot for user in PSP is because it is controlling a corresponding field in the pod spec

liggitt commented on pkg/security/podsecuritypolicy/group/types.go:
```diff
@@ -31,5 +31,5 @@ type GroupStrategy interface {
 	GenerateSingle(pod *api.Pod) (*int64, error)
 	// Validate ensures that the specified values fall within the range of the strategy.
-	Validate(fldPath *field.Path, pod *api.Pod, groups []int64) field.ErrorList
+	Validate(fldPath *field.Path, pod *api.Pod, runAsNonRoot *bool, groups []int64) field.ErrorList
```
defer the runAsNonRoot aspects until https://github.com/kubernetes/kubernetes/pull/62216 is settled?

liggitt commented on pkg/security/podsecuritypolicy/factory.go:
```diff
@@ -111,6 +117,20 @@ func createUserStrategy(opts *policy.RunAsUserStrategyOptions) (user.RunAsUserSt
+	case policy.RunAsGroupStrategyRunAsAny:
+		return group.NewRunAsAny()
+	default:
+		return nil, fmt.Errorf("Unrecognized RunAsGroup strategy type %s", opts.Rule)
```
doesn't this mean a PSP has to specify a runasgroup strategy? we cannot require this field, since it's a new field. what is the default going to be?

liggitt commented on staging/src/k8s.io/api/extensions/v1beta1/types.go:
```diff
@@ -965,6 +965,8 @@ type PodSecurityPolicySpec struct {
 	// +optional
 	ForbiddenSysctls []string `json:"forbiddenSysctls,omitempty" protobuf:"bytes,20,rep,name=forbiddenSysctls"`
+	// runAsGroup is the strategy that will dictate the allowable RunAsGroup values that may be set.
+	RunAsGroup RunAsGroupStrategyOptions `json:"runAsGroup" protobuf:"bytes,21,opt,name=runAsGroup"`
```
because this is a new field, this has to be optional, which means it must also be a pointer, and marked omitempty (comment applies to other types files as well)

liggitt replied:
we also need to determine what the default behavior is when this is nil

tallclair replied:
The default behavior must be "RunAsAny", otherwise existing policies will start rejecting pods that were previously allowed.

tallclair commented on staging/src/k8s.io/api/extensions/v1beta1/types.go:
```diff
@@ -1067,6 +1069,17 @@ type RunAsUserStrategyOptions struct {
 }
 
+// RunAsGroupStrategyOptions defines the strategy type and any options used to create the strategy.
+// Deprecated: use RunAsGroupStrategyOptions from policy API Group instead.
```
@liggitt do you know how this works as a multi-group API? Do we need to add it here, or can we just add new fields to the policy group API?

liggitt replied:
yeah, needs to be added here for round trip support

tallclair commented on pkg/security/podsecuritypolicy/group/nonroot.go:
```diff
@@ -0,0 +1,68 @@
+func (s *nonRoot) Validate(scPath *field.Path, _ *api.Pod, runAsNonRoot *bool, groups []int64) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if runAsNonRoot == nil && len(groups) == 0 {
```
Why are you checking that a group is specified here?

tallclair commented on pkg/security/podsecuritypolicy/group/nonroot.go:
```diff
@@ -0,0 +1,68 @@
+	if runAsNonRoot != nil && *runAsNonRoot == false {
+		allErrs = append(allErrs, field.Invalid(scPath.Child("runAsNonRoot"), *runAsNonRoot, "must be true"))
+		return allErrs
+	}
```
It seems a little weird to me that RunAsNonRoot group checks the value of runAsNonRoot (user). I think I understand the motivation, but minimally this should accept manually setting a non-root user too. Still, I'm not sure that belongs in the group strategy. If we're going down that road, why not check supplemental & fs groups too?

tallclair commented on pkg/security/podsecuritypolicy/group/nonroot.go:
```diff
@@ -0,0 +1,68 @@
+		allErrs = append(allErrs, field.Invalid(scPath.Child("runAsNonRoot"), *runAsNonRoot, "must be true"))
+		return allErrs
+	}
+	for _, group := range groups {
```
IMO this should require that there is a group. Actually, it's a little weird for supplemental & fs groups.

liggitt commented on pkg/apis/policy/types.go:
```diff
@@ -356,6 +369,17 @@ const (
+type RunAsGroupStrategy string
+
+const (
+	// RunAsGroupStrategyMustRunAs means that container must run as a particular uid.
```
s/uid/gid/?

liggitt commented on pkg/apis/policy/types.go:
```diff
@@ -356,6 +369,17 @@ const (
+const (
+	// RunAsGroupStrategyMustRunAs means that container must run as a particular uid.
+	RunAsGroupStrategyMustRunAs RunAsGroupStrategy = "MustRunAs"
+	// RunAsGroupStrategyRunAsAny means that container may make requests for any uid.
```
gid

liggitt commented on pkg/apis/policy/v1beta1/defaults.go:
```diff
@@ -32,4 +32,8 @@ func SetDefaults_PodSecurityPolicySpec(obj *policyv1beta1.PodSecurityPolicySpec)
 		obj.AllowPrivilegeEscalation = &t
 	}
+	if obj.RunAsGroup == nil {
+		runAsGroupOpts := policyv1beta1.RunAsGroupStrategyOptions{Rule: policyv1beta1.RunAsGroupStrategyRunAsAny}
```
no defaulting on auto-cleared alpha fields. this causes PSP objects read from the API to contain this field, which we don't want when the alpha feature is disabled

liggitt commented on pkg/security/podsecuritypolicy/factory.go:
```diff
@@ -111,6 +117,18 @@ func createUserStrategy(opts *policy.RunAsUserStrategyOptions) (user.RunAsUserSt
 }
 
+// createGroupStrategy creates a new group strategy.
+func createGroupStrategy(opts *policy.RunAsGroupStrategyOptions) (group.GroupStrategy, error) {
```
need to handle the case where no RunAsGroup policy was specified

liggitt commented on pkg/security/podsecuritypolicy/factory.go:
```diff
@@ -47,6 +47,11 @@ func (f *simpleStrategyFactory) CreateStrategies(psp *policy.PodSecurityPolicy,
 		errs = append(errs, err)
 	}
 
+	groupStrat, err := createGroupStrategy(psp.Spec.RunAsGroup)
```
I expected this omitted if the feature gate was off (leave `groupStrat` as nil if the feature is disabled)

liggitt commented on pkg/security/podsecuritypolicy/provider.go:
```diff
@@ -121,6 +121,14 @@ func (s *simpleProvider) DefaultContainerSecurityContext(pod *api.Pod, container
 		sc.SetRunAsUser(uid)
 	}
 
+	if sc.RunAsGroup() == nil {
```
protect this whole block with the feature gate check

liggitt commented on pkg/security/podsecuritypolicy/provider.go:
```diff
@@ -280,6 +288,12 @@ func (s *simpleProvider) ValidateContainer(pod *api.Pod, container *api.Containe
+	if sc.RunAsGroup() != nil {
+		runAsGroups = []int64{*sc.RunAsGroup()}
+	}
+	allErrs = append(allErrs, s.strategies.RunAsGroupStrategy.Validate(scPath, pod, runAsGroups)...)
```
protect this with the feature gate check

liggitt commented on staging/src/k8s.io/api/policy/v1beta1/types.go:
```diff
@@ -174,6 +174,9 @@ type PodSecurityPolicySpec struct {
 	// runAsUser is the strategy that will dictate the allowable RunAsUser values that may be set.
 	RunAsUser RunAsUserStrategyOptions `json:"runAsUser" protobuf:"bytes,11,opt,name=runAsUser"`
+	// RunAsGroup is the strategy that will dictate the allowable RunAsGroup values that may be set.
+	// This field requires the RunAsGroup alpha feature gate to be enabled.
```
add `+optional`

liggitt commented on staging/src/k8s.io/api/extensions/v1beta1/types.go:
```diff
@@ -918,6 +918,9 @@ type PodSecurityPolicySpec struct {
 	// runAsUser is the strategy that will dictate the allowable RunAsUser values that may be set.
 	RunAsUser RunAsUserStrategyOptions `json:"runAsUser" protobuf:"bytes,11,opt,name=runAsUser"`
+	// RunAsGroup is the strategy that will dictate the allowable RunAsGroup values that may be set.
+	// This field requires the RunAsGroup alpha feature gate to be enabled.
```
add `+optional`

liggitt commented on test/integration/dryrun/dryrun_test.go:
```diff
@@ -212,7 +212,7 @@ var dryrunData = map[schema.GroupVersionResource]struct {
 	},
 	gvr("extensions", "v1beta1", "podsecuritypolicies"): {
-		stub: `{"metadata": {"name": "psp1"}, "spec": {"fsGroup": {"rule": "RunAsAny"}, "privileged": true, "runAsUser": {"rule": "RunAsAny"}, "seLinux": {"rule": "MustRunAs"}, "supplementalGroups": {"rule": "RunAsAny"}}}`,
+		stub: `{"metadata": {"name": "psp1"}, "spec": {"fsGroup": {"rule": "RunAsAny"}, "privileged": true, "runAsUser": {"rule": "RunAsAny"}, "runAsGroup": {"rule": "RunAsAny"}, "seLinux": {"rule": "MustRunAs"}, "supplementalGroups": {"rule": "RunAsAny"}}}`,
```
shouldn't need to update these tests yet, these stubs don't need to exercise every field, just be the minimal valid object

liggitt commented on test/integration/etcd/etcd_storage_path_test.go:
```diff
@@ -269,7 +269,7 @@ var etcdStorageData = map[schema.GroupVersionResource]struct {
 	},
 	gvr("extensions", "v1beta1", "podsecuritypolicies"): {
-		stub:             `{"metadata": {"name": "psp1"}, "spec": {"fsGroup": {"rule": "RunAsAny"}, "privileged": true, "runAsUser": {"rule": "RunAsAny"}, "seLinux": {"rule": "MustRunAs"}, "supplementalGroups": {"rule": "RunAsAny"}}}`,
+		stub:             `{"metadata": {"name": "psp1"}, "spec": {"fsGroup": {"rule": "RunAsAny"}, "privileged": true, "runAsUser": {"rule": "RunAsAny"}, "runAsGroup": {"rule": "RunAsAny"}, "seLinux": {"rule": "MustRunAs"}, "supplementalGroups": {"rule": "RunAsAny"}}}`,
```
same here, shouldn't need to update these tests yet, these stubs don't need to exercise every field, just be the minimal valid object

liggitt commented on pkg/securitycontext/accessors.go:
```diff
@@ -29,6 +29,7 @@ type PodSecurityContextAccessor interface {
 	HostIPC() bool
 	SELinuxOptions() *api.SELinuxOptions
 	RunAsUser() *int64
+	RunAsGroup() *int64
```
test in TestPodSecurityContextAccessor

liggitt commented on pkg/securitycontext/accessors.go:
```diff
@@ -142,6 +143,20 @@ func (w *podSecurityContextWrapper) SetRunAsUser(v *int64) {
+	}
+	return w.podSC.RunAsGroup
+}
+func (w *podSecurityContextWrapper) SetRunAsGroup(v *int64) {
```
add to PodSecurityContextMutator and test in TestPodSecurityContextMutator

liggitt commented on pkg/securitycontext/accessors.go:
```diff
@@ -205,6 +221,7 @@ type ContainerSecurityContextMutator interface {
 	SetPrivileged(*bool)
 	SetSELinuxOptions(*api.SELinuxOptions)
 	SetRunAsUser(*int64)
+	SetRunAsGroup(*int64)
```
test in TestContainerSecurityContextMutator and TestEffectiveContainerSecurityContextMutator

liggitt commented on pkg/securitycontext/accessors.go:
```diff
@@ -191,6 +206,7 @@ type ContainerSecurityContextAccessor interface {
 	ProcMount() api.ProcMountType
 	SELinuxOptions() *api.SELinuxOptions
 	RunAsUser() *int64
+	RunAsGroup() *int64
```
test in TestContainerSecurityContextAccessor and TestEffectiveContainerSecurityContextAccessor

liggitt commented on staging/src/k8s.io/api/extensions/v1beta1/types.go:
```diff
@@ -918,6 +918,10 @@ type PodSecurityPolicySpec struct {
 	// runAsUser is the strategy that will dictate the allowable RunAsUser values that may be set.
 	RunAsUser RunAsUserStrategyOptions `json:"runAsUser" protobuf:"bytes,11,opt,name=runAsUser"`
+	// RunAsGroup is the strategy that will dictate the allowable RunAsGroup values that may be set.
+	// This field requires the RunAsGroup alpha feature gate to be enabled.
```
indicate any runAsGroup value is allowed when this field is omitted

liggitt commented on staging/src/k8s.io/api/policy/v1beta1/types.go:
```diff
@@ -174,6 +174,10 @@ type PodSecurityPolicySpec struct {
 	// runAsUser is the strategy that will dictate the allowable RunAsUser values that may be set.
 	RunAsUser RunAsUserStrategyOptions `json:"runAsUser" protobuf:"bytes,11,opt,name=runAsUser"`
+	// RunAsGroup is the strategy that will dictate the allowable RunAsGroup values that may be set.
+	// This field requires the RunAsGroup alpha feature gate to be enabled.
```
indicate any runAsGroup value is allowed when this field is omitted

liggitt commented on staging/src/k8s.io/api/policy/v1beta1/types.go:
```diff
@@ -319,6 +323,16 @@ type RunAsUserStrategyOptions struct {
+type RunAsGroupStrategyOptions struct {
+	// rule is the strategy that will dictate the allowable RunAsUser values that may be set.
+	Rule RunAsGroupStrategy `json:"rule" protobuf:"bytes,1,opt,name=rule,casttype=RunAsGroupStrategy"`
+	// ranges are the allowed ranges of uids that may be used. If you would like to force a single uid
```
gids, gid

liggitt commented on staging/src/k8s.io/api/policy/v1beta1/types.go:
```diff
@@ -319,6 +323,16 @@ type RunAsUserStrategyOptions struct {
 
+// RunAsGroupStrategyOptions defines the strategy type and any options used to create the strategy.
+type RunAsGroupStrategyOptions struct {
+	// rule is the strategy that will dictate the allowable RunAsUser values that may be set.
```
RunAsGroup

liggitt commented on pkg/apis/policy/validation/validation.go:
```diff
@@ -233,6 +234,38 @@ func validatePSPRunAsUser(fldPath *field.Path, runAsUser *policy.RunAsUserStrate
+		}
+	case policy.RunAsGroupStrategyMustRunAs:
+		if len(runAsGroup.Ranges) < 1 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("ranges"), runAsGroup.Ranges, "Must provide at least one 'Ranges'"))
```
"must provide at least one range"

liggitt commented on pkg/apis/policy/validation/validation.go:
```diff
@@ -233,6 +234,38 @@ func validatePSPRunAsUser(fldPath *field.Path, runAsUser *policy.RunAsUserStrate
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("ranges"), runAsGroup.Ranges, "Ranges must be empty"))
+		}
+	case policy.RunAsGroupStrategyMustRunAs:
+		if len(runAsGroup.Ranges) < 1 {
```
`if len(...) == 0`

liggitt commented on pkg/apis/policy/types.go:
```diff
@@ -182,6 +182,9 @@ type PodSecurityPolicySpec struct {
 	// RunAsUser is the strategy that will dictate the allowable RunAsUser values that may be set.
 	RunAsUser RunAsUserStrategyOptions
+	// RunAsGroup is the strategy that will dictate the allowable RunAsGroup values that may be set.
+	// This field requires the RunAsGroup alpha feature gate to be enabled.
```
indicate any runAsGroup value is allowed when this field is omitted

liggitt reviewed (commented):
did a last sweep through, found a few doc updates and missing tests

lgtm otherwise. will need an exception if you want to target this for 1.12. I don't feel strongly either way about it (seems low-risk, but also not high priority to get in in alpha state). @tallclair, thoughts either way?

krmayankk commented:
@liggitt thanks for quick reviews. Addressed all comments. I will file an exception request. If this is low risk and alpha, please consider it for this release. /cc  @tallclair

krmayankk commented:
/test pull-kubernetes-verify

liggitt commented:
`pkg/controller/apis/config/v1alpha1/zz_generated.conversion.go` is a spurious change and is failing validation. you might need to rebase on master and/or regenerate conversions

krmayankk commented:
@liggitt  fixed

tallclair commented on pkg/apis/policy/fuzzer/fuzzer.go:
```diff
@@ -40,6 +40,12 @@ var Funcs = func(codecs runtimeserializer.CodecFactory) []interface{} {
 			}
 			psp.RunAsUser.Rule = runAsUserRules[c.Rand.Intn(len(runAsUserRules))]
 
+			runAsGroupRules := []policy.RunAsGroupStrategy{
```
note: [MayRunAs](https://github.com/kubernetes/kubernetes/pull/65135) should be added here once that PR merges

tallclair replied:
It merged. I think you should update this PR to include that strategy.

krmayankk replied:
fixed

tallclair commented on pkg/apis/policy/types.go:
```diff
@@ -182,6 +182,10 @@ type PodSecurityPolicySpec struct {
 	RunAsUser RunAsUserStrategyOptions
+	// RunAsGroup is the strategy that will dictate the allowable RunAsGroup values that may be set.
+	// If this field is omitted, RunAsGroup can take any value. This field requires the RunAsGroup
+	// alpha feature gate to be enabled.
```
nit: drop "alpha" in this sentence.

tallclair replied:
Also need to update the comment in staging/src/k8s.io/api/policy/v1beta1/types.go and (maybe) staging/src/k8s.io/api/extensions/v1beta1/types.go

krmayankk replied:
fixed both files @tallclair

tallclair commented on pkg/apis/policy/types.go:
```diff
@@ -182,6 +182,10 @@ type PodSecurityPolicySpec struct {
 	// RunAsUser is the strategy that will dictate the allowable RunAsUser values that may be set.
 	RunAsUser RunAsUserStrategyOptions
+	// RunAsGroup is the strategy that will dictate the allowable RunAsGroup values that may be set.
+	// If this field is omitted, RunAsGroup can take any value. This field requires the RunAsGroup
```
nit: clarify "If this field is omitted, **the pod's** RunAsGroup can take any value."

tallclair replied:
Also need to update the comment in staging/src/k8s.io/api/policy/v1beta1/types.go and (maybe) staging/src/k8s.io/api/extensions/v1beta1/types.go

tallclair commented on pkg/apis/policy/validation/validation.go:
```diff
@@ -233,6 +234,38 @@ func validatePSPRunAsUser(fldPath *field.Path, runAsUser *policy.RunAsUserStrate
 
+// validatePSPRunAsGroup validates the RunAsGroup fields of PodSecurityPolicy.
+func validatePSPRunAsGroup(fldPath *field.Path, runAsGroup *policy.RunAsGroupStrategyOptions) field.ErrorList {
+	allErrs := field.ErrorList{}
```
nit: prefer `var allErrs field.ErrorList`

tallclair commented on pkg/apis/policy/validation/validation.go:
```diff
@@ -233,6 +234,38 @@ func validatePSPRunAsUser(fldPath *field.Path, runAsUser *policy.RunAsUserStrate
+	supportedRunAsGroupRules := sets.NewString(
+		string(policy.RunAsGroupStrategyMustRunAs),
+		string(policy.RunAsGroupStrategyRunAsAny),
+	)
```
nit: Use a slice `[]RunAsGroupStrategy`, and move to where it's used.

tallclair commented on pkg/security/podsecuritypolicy/factory.go:
```diff
@@ -111,6 +123,21 @@ func createUserStrategy(opts *policy.RunAsUserStrategyOptions) (user.RunAsUserSt
 }
 
+// createGroupStrategy creates a new group strategy.
+func createGroupStrategy(opts *policy.RunAsGroupStrategyOptions) (group.GroupStrategy, error) {
```
nit: rename `createRunAsGroupStrategy`

tallclair commented on pkg/security/podsecuritypolicy/group/mustrunas_test.go:
```diff
@@ -113,6 +113,7 @@ func TestValidate(t *testing.T) {
 	tests := map[string]struct {
 		ranges        []policy.IDRange
 		groups        []int64
+		nonRoot       bool
```
revert this file

tallclair reviewed (commented):
mostly nits

krmayankk commented:
@tallclair  addressed all comments

tallclair commented on staging/src/k8s.io/api/extensions/v1beta1/generated.proto:
```diff
@@ -847,6 +847,12 @@ message PodSecurityPolicySpec {
+  // alpha feature gate to be enabled.
+  // +optional
+  optional RunAsGroupStrategyOptions runAsGroup = 22;
+
```
@liggitt This version of PSP is deprecated (moved to the policy API group). Do we still need to add new fields here, or can we let this version fall behind?

liggitt replied:
we need to keep field parity until we are sure we will never have to convert from policy/v1beta1 to extensions/v1beta1. That will occur once we remove the ability to influence etcd storage versions and stop serving PSPs under the extensions/v1beta1 API group. That may happen in 1.13, but I'm not prepared to commit to that at this point.

liggitt commented on pkg/apis/policy/validation/validation.go:
```diff
@@ -233,6 +234,44 @@ func validatePSPRunAsUser(fldPath *field.Path, runAsUser *policy.RunAsUserStrate
+			allErrs = append(allErrs, validateGroupIDRange(fldPath.Child("ranges").Index(idx), rng)...)
+		}
+	case policy.RunAsGroupStrategyMayRunAs:
+		// validate range settings
```
are ranges not required for this strategy?

krmayankk replied:
yes they are not required. If the  Ranges are specified, then RUnAsGroup should be in that range, else ignored @liggitt

krmayankk replied:
sorry i misunderstood that. I believe this check should be added to FSGroup and SupplementalGroup as well ? I  will fix that as part of a separate PR @tallclair @liggitt

krmayankk commented on pkg/security/podsecuritypolicy/factory.go:
```diff
@@ -111,6 +123,21 @@ func createUserStrategy(opts *policy.RunAsUserStrategyOptions) (user.RunAsUserSt
+	case policy.RunAsGroupStrategyMustRunAs:
+		return group.NewMustRunAs(opts.Ranges)
+	case policy.RunAsGroupStrategyRunAsAny:
+		return group.NewRunAsAny()
```
fixed

tallclair commented:
/lgtm

krmayankk commented:
/assign @smarterclayton

smarterclayton commented:
/approve
